src/rabbit_queue_type.erl

```diff
@@ -0,0 +1,491 @@
+-type action() ::
+    {monitor, Pid :: pid(), queue_ref()} |
+    {link_name, queue_ref(), [queue_ref()]} |
+    {deliver, rabbit_type:ctag(), boolean(), [rabbit_amqqueue:qmsg()]}.
+
+-type actions() :: [action()].
+
+-record(ctx, {module :: module(),
+              name :: queue_name(),
+              state :: queue_state()}).
+
+-opaque ctxs() :: #{queue_ref() => #ctx{} | queue_ref()}.
+
+-type consume_spec() :: #{no_ack := boolean(),
+                          channel_pid := pid(),
+                          limiter_pid => pid(),
+                          limiter_active => boolean(),
+                          prefetch_count => non_neg_integer(),
+                          consumer_tag := rabbit_types:ctag(),
+                          exclusive_consume => boolean(),
+                          args => rabbit_framing:amqp_table(),
+                          ok_msg := term(),
+                          acting_user :=  rabbit_types:username()}.
+
+
+-export_type([ctxs/0,
+              consume_spec/0,
+              action/0,
+              actions/0]).
+
+
+% copied from rabbit_amqqueue
+-type absent_reason() :: 'nodedown' | 'crashed' | stopped | timeout.
+
+%% is the queue type feature enabled
+-callback is_enabled() -> boolean().
+
+-callback declare(amqqueue:amqqueue(), node()) ->
+    {'new' | 'existing' | 'owner_died', amqqueue:amqqueue()} |
+    {'absent', amqqueue:amqqueue(), absent_reason()} |
+    rabbit_types:channel_exit().
+
+-callback delete(amqqueue:amqqueue(),
+                 boolean(),
+                 boolean(),
+                 rabbit_types:username()) ->
+    rabbit_types:ok(non_neg_integer()) |
+    rabbit_types:error(in_use | not_empty).
+
+-callback recover(rabbit_types:vhost(), [amqqueue:amqqueue()]) ->
+    {Recovered :: [amqqueue:amqqueue()],
+     Failed :: [amqqueue:amqqueue()]}.
+
+%% checks if the queue should be recovered
+-callback is_recoverable(amqqueue:amqqueue()) ->
+    boolean().
+
+-callback purge(amqqueue:amqqueue()) ->
+    {ok, non_neg_integer()} | {error, term()}.
+
+-callback policy_changed(amqqueue:amqqueue()) -> ok.
+
+%% stateful
+%% intitialise and return a queue type specific session context
+-callback init(amqqueue:amqqueue()) -> queue_state().
+
+-callback consume(amqqueue:amqqueue(),
+                  consume_spec(),
+                  queue_state()) ->
+    {ok, queue_state(), actions()} | {error, term()}.
+
+-callback cancel(amqqueue:amqqueue(),
+                 rabbit_types:ctag(),
+                 term(),
+                 rabbit_types:username(),
+                 queue_state()) ->
+    {ok, queue_state(), actions()} | {error, term()}.
+
+%% any async events returned from the queue system should be processed through
+%% this
+-callback handle_event(Event :: term(),
+                       queue_state()) ->
+    {ok, queue_state(), actions()} | {error, term()} | eol.
+
+-callback deliver([{amqqueue:amqqueue(), queue_state()}],
```

Comment:
After a bit more thoughts, I find this callback (and all others working on a list of queues) a bit more weird in fact compared to last week:
1. As I said, from an API PoV, this looks inconsistent: some callbacks take a list of queues, others a single queue, and the name doesn't reflect this change of behavior.
2. Should the queues in that list share the same queue type? What if a queue type implementation finds a queue from a different queue type in the middle of that list?

It seems to me that this feature should be outside of the new behavior and it should call an actual `deliver()` callback which takes a queue as argument. Again, my comment applies to other callbacks such as `recover()`.